In [1]:
# import the library
%matplotlib inline

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os

# Part 1: Load the Weather data

In [2]:
df_c = pd.read_csv('../../Data/Raw/city_attributes.csv')
df_h = pd.read_csv('../../Data/Raw/humidity.csv')
df_p = pd.read_csv('../../Data/Raw/pressure.csv')
df_t = pd.read_csv('../../Data/Raw/temperature.csv')
df_d = pd.read_csv('../../Data/Raw/weather_description.csv')
df_wd = pd.read_csv('../../Data/Raw/wind_direction.csv')
df_ws = pd.read_csv('../../Data/Raw/wind_speed.csv')

In [3]:
#size and number of variables in the dataset
print(df_c.shape)
print(df_h.shape)
print(df_p.shape)
print(df_t.shape)
print(df_d.shape)
print(df_wd.shape)
print(df_ws.shape)

(36, 4)
(45253, 37)
(45253, 37)
(45253, 37)
(45253, 37)
(45253, 37)
(45253, 37)


In [4]:
#Let us look at the columns of the city_attributes.csv
print(df_c.dtypes)

City          object
Country       object
Latitude     float64
Longitude    float64
dtype: object


In [5]:
#Let us look at the columns of the other csv: humidty, pressure, temperature, wind_direction, wind_speed
print(df_h.dtypes)
print(df_p.dtypes)
print(df_t.dtypes)
print(df_wd.dtypes)
print(df_ws.dtypes)

datetime              object
Vancouver            float64
Portland             float64
San Francisco        float64
Seattle              float64
Los Angeles          float64
San Diego            float64
Las Vegas            float64
Phoenix              float64
Albuquerque          float64
Denver               float64
San Antonio          float64
Dallas               float64
Houston              float64
Kansas City          float64
Minneapolis          float64
Saint Louis          float64
Chicago              float64
Nashville            float64
Indianapolis         float64
Atlanta              float64
Detroit              float64
Jacksonville         float64
Charlotte            float64
Miami                float64
Pittsburgh           float64
Toronto              float64
Philadelphia         float64
New York             float64
Montreal             float64
Boston               float64
Beersheba            float64
Tel Aviv District    float64
Eilat                float64
Haifa         

In [6]:
#Let us look at the columns of the weather_description.csv
print(df_d.dtypes)
#Note: all columns are defined as objects

datetime             object
Vancouver            object
Portland             object
San Francisco        object
Seattle              object
Los Angeles          object
San Diego            object
Las Vegas            object
Phoenix              object
Albuquerque          object
Denver               object
San Antonio          object
Dallas               object
Houston              object
Kansas City          object
Minneapolis          object
Saint Louis          object
Chicago              object
Nashville            object
Indianapolis         object
Atlanta              object
Detroit              object
Jacksonville         object
Charlotte            object
Miami                object
Pittsburgh           object
Toronto              object
Philadelphia         object
New York             object
Montreal             object
Boston               object
Beersheba            object
Tel Aviv District    object
Eilat                object
Haifa                object
Nahariyya           

In [7]:
#let us look at details of numeric variables
print(df_c.describe())
print(df_h.describe())
print(df_p.describe())
print(df_t.describe())
print(df_wd.describe())
print(df_ws.describe())

        Latitude   Longitude
count  36.000000   36.000000
mean   37.066743  -73.544668
std     5.815514   51.612349
min    25.774269 -123.119339
25%    32.766126 -105.401312
50%    36.170429  -86.471241
75%    40.998211  -74.874332
max    49.249660   35.216331
          Vancouver      Portland  San Francisco       Seattle   Los Angeles  \
count  43427.000000  44804.000000   44311.000000  44964.000000  45101.000000   
mean      81.895480     74.697616      76.875042     77.159038     62.773841   
std       14.522221     19.042656      17.396016     18.147464     21.818042   
min       12.000000     10.000000       6.000000     13.000000      5.000000   
25%       73.000000     63.000000      68.000000     66.000000     48.000000   
50%       86.000000     80.000000      81.000000     81.000000     66.000000   
75%       93.000000     90.000000      89.000000     93.000000     81.000000   
max      100.000000    100.000000     100.000000    100.000000    100.000000   

          San Dieg

          Vancouver      Portland  San Francisco       Seattle   Los Angeles  \
count  44458.000000  45252.000000   44459.000000  45252.000000  45252.000000   
mean     159.892978    189.900071     209.299489    161.380226    139.549611   
std       98.163221    102.873761      95.589055    103.206816    105.591486   
min        0.000000      0.000000       0.000000      0.000000      0.000000   
25%       80.000000    107.000000     160.000000     70.000000     29.000000   
50%      140.000000    198.000000     232.000000    175.000000    140.000000   
75%      236.000000    289.000000     280.000000    218.000000    230.000000   
max      360.000000    360.000000     360.000000    360.000000    360.000000   

          San Diego     Las Vegas       Phoenix   Albuquerque        Denver  \
count  45252.000000  45248.000000  45252.000000  45252.000000  45252.000000   
mean     199.212852    176.320987    168.564351    199.878458    184.769005   
std      106.350252    106.902903    107.5

In [8]:
c = df_c.dtypes.index
for x in c:
    missing=np.where(df_c[x].isnull()==True)
    print(x, "column has", len(missing[0]), "missing values")

City column has 0 missing values
Country column has 0 missing values
Latitude column has 0 missing values
Longitude column has 0 missing values


In [9]:
# identify cities with more than 700 missing humidity values
c = df_h.dtypes.index
for x in c:
    missing=np.where(df_h[x].isnull()==True)
    if len(missing[0]) >= 700: 
        print(x, "column has", len(missing[0]), "missing values")

Vancouver column has 1826 missing values
San Francisco column has 942 missing values
Las Vegas column has 842 missing values
Phoenix column has 1308 missing values
Albuquerque column has 710 missing values
Denver column has 1808 missing values
Saint Louis column has 1289 missing values
Chicago column has 1109 missing values
Detroit column has 862 missing values
Miami column has 1087 missing values
Toronto column has 728 missing values
New York column has 1624 missing values
Montreal column has 1696 missing values
Beersheba column has 859 missing values
Tel Aviv District column has 1113 missing values
Eilat column has 970 missing values
Haifa column has 818 missing values
Nahariyya column has 817 missing values
Jerusalem column has 906 missing values


In [10]:
# identify cities with more than 700 missing pressure values
c = df_p.dtypes.index
for x in c:
    missing=np.where(df_p[x].isnull()==True)
    if len(missing[0]) >= 700: 
        print(x, "column has", len(missing[0]), "missing values")

Vancouver column has 4234 missing values
San Francisco column has 815 missing values
Chicago column has 725 missing values
Miami column has 794 missing values
New York column has 1041 missing values
Montreal column has 994 missing values
Beersheba column has 806 missing values
Tel Aviv District column has 930 missing values
Eilat column has 792 missing values
Haifa column has 798 missing values
Nahariyya column has 795 missing values
Jerusalem column has 799 missing values


In [11]:
# identify cities with more than 700 missing temperature values
c = df_t.dtypes.index
for x in c:
    missing=np.where(df_t[x].isnull()==True)
    if len(missing[0]) >= 700: 
        print(x, "column has", len(missing[0]), "missing values")

Vancouver column has 795 missing values
San Francisco column has 793 missing values
Miami column has 805 missing values
New York column has 793 missing values
Beersheba column has 798 missing values
Tel Aviv District column has 793 missing values
Eilat column has 792 missing values
Haifa column has 798 missing values
Nahariyya column has 797 missing values
Jerusalem column has 793 missing values


In [12]:
# identify cities with more than 700 missing description values
c = df_d.dtypes.index
for x in c:
    missing=np.where(df_d[x].isnull()==True)
    if len(missing[0]) >= 700: 
        print(x, "column has", len(missing[0]), "missing values")

Vancouver column has 793 missing values
San Francisco column has 793 missing values
Miami column has 793 missing values
New York column has 793 missing values
Beersheba column has 793 missing values
Tel Aviv District column has 793 missing values
Eilat column has 792 missing values
Haifa column has 793 missing values
Nahariyya column has 793 missing values
Jerusalem column has 793 missing values


In [13]:
# identify cities with more than 700 missing wind direction values
c = df_wd.dtypes.index
for x in c:
    missing=np.where(df_wd[x].isnull()==True)
    if len(missing[0]) >= 700: 
        print(x, "column has", len(missing[0]), "missing values")

Vancouver column has 795 missing values
San Francisco column has 794 missing values
Miami column has 795 missing values
New York column has 793 missing values
Beersheba column has 793 missing values
Tel Aviv District column has 793 missing values
Eilat column has 797 missing values
Haifa column has 793 missing values
Nahariyya column has 793 missing values
Jerusalem column has 793 missing values


In [14]:
# identify cities with more than 700 missing wind speed values
c = df_ws.dtypes.index
for x in c:
    missing=np.where(df_ws[x].isnull()==True)
    if len(missing[0]) >= 700: 
        print(x, "column has", len(missing[0]), "missing values")

Vancouver column has 795 missing values
San Francisco column has 794 missing values
Miami column has 795 missing values
New York column has 793 missing values
Beersheba column has 793 missing values
Tel Aviv District column has 794 missing values
Eilat column has 792 missing values
Haifa column has 795 missing values
Nahariyya column has 794 missing values
Jerusalem column has 794 missing values


In [15]:
c = df_d.dtypes.index
for x in c:
    print(x)
    print(df_d[x].unique())

datetime
['2012-10-01 12:00:00' '2012-10-01 13:00:00' '2012-10-01 14:00:00' ...
 '2017-11-29 22:00:00' '2017-11-29 23:00:00' '2017-11-30 00:00:00']
Vancouver
[nan 'mist' 'broken clouds' 'sky is clear' 'light rain' 'few clouds' 'fog'
 'overcast clouds' 'light intensity shower rain' 'moderate rain'
 'light intensity drizzle' 'scattered clouds' 'proximity shower rain'
 'heavy intensity rain' 'heavy snow' 'shower rain' 'snow'
 'heavy shower snow' 'light intensity drizzle rain' 'light snow'
 'very heavy rain' 'smoke' 'thunderstorm with heavy rain'
 'light shower snow' 'thunderstorm' 'thunderstorm with light rain' 'haze'
 'dust' 'volcanic ash' 'heavy intensity shower rain'
 'thunderstorm with rain' 'sleet' 'light rain and snow' 'drizzle'
 'shower snow' 'light shower sleet' 'proximity thunderstorm'
 'ragged thunderstorm']
Portland
[nan 'scattered clouds' 'sky is clear' 'few clouds' 'broken clouds'
 'overcast clouds' 'mist' 'haze' 'dust' 'fog' 'moderate rain' 'light rain'
 'heavy intensity rai

[nan 'overcast clouds' 'broken clouds' 'light rain' 'moderate rain'
 'light intensity drizzle' 'heavy intensity rain' 'mist' 'drizzle'
 'scattered clouds' 'few clouds' 'sky is clear' 'thunderstorm'
 'proximity thunderstorm' 'thunderstorm with rain'
 'heavy intensity drizzle' 'fog' 'haze' 'heavy snow' 'snow'
 'thunderstorm with light rain' 'freezing rain' 'light snow'
 'proximity thunderstorm with rain' 'thunderstorm with heavy rain'
 'very heavy rain' 'squalls' 'thunderstorm with light drizzle'
 'light intensity shower rain']
Atlanta
[nan 'light rain' 'overcast clouds' 'heavy intensity rain' 'moderate rain'
 'mist' 'light intensity drizzle' 'scattered clouds' 'broken clouds'
 'drizzle' 'sky is clear' 'fog' 'few clouds' 'haze'
 'thunderstorm with rain' 'thunderstorm with light rain' 'heavy snow'
 'proximity thunderstorm' 'heavy intensity drizzle'
 'light intensity shower rain' 'thunderstorm'
 'proximity thunderstorm with rain' 'thunderstorm with heavy rain'
 'squalls' 'thunderstorm with

# Conclusion: 

There are 36 cities identified in the 7 csv files;  
6 files contains weather information of the 36 cities;  
All csv files contains numerical information, except weather_description, which is textual.   

In each of the 6 files:  
-datetime is defined as an object  
-many missing values for some of the cities  

Relative humidity is measured in percentage(%); 
Pressure is measured in hPa???;  
Temperatire is measured in Kelvin;  
Wind direction is measured in meteorological degrees;  
Wind speed is measured in mps??



# Part 2: Load the 2017 bixi data

In [16]:
df_bixi = pd.read_csv('../../Data/Raw/MTL_data/OD_2017.csv',low_memory=False)
df_stn = pd.read_csv('../../Data/Raw/MTL_data/Stations_2017.csv',low_memory=False)

In [17]:
#size and number of variables in the OD csv
print(df_bixi.shape)
print(df_bixi.head(10))

(4018722, 7)
   Unnamed: 0        start_date  start_station_code          end_date  \
0           0  2017-04-15 00:00                7060  2017-04-15 00:31   
1           1  2017-04-15 00:01                6173  2017-04-15 00:10   
2           2  2017-04-15 00:01                6203  2017-04-15 00:04   
3           3  2017-04-15 00:01                6104  2017-04-15 00:06   
4           4  2017-04-15 00:01                6174  2017-04-15 00:11   
5           5  2017-04-15 00:02                6719  2017-04-15 00:12   
6           6  2017-04-15 00:02                6223  2017-04-15 00:13   
7           7  2017-04-15 00:02                6253  2017-04-15 00:07   
8           8  2017-04-15 00:02                6119  2017-04-15 00:38   
9           9  2017-04-15 00:02                6119  2017-04-15 00:37   

  end_station_code  duration_sec  is_member  
0             7060          1841          1  
1             6173           553          1  
2             6204           195          1  

In [18]:
#Let us look at the columns of OD
print(df_bixi.dtypes)
#Let us look at details of numeric variables
print(df_bixi.describe())

Unnamed: 0             int64
start_date            object
start_station_code     int64
end_date              object
end_station_code      object
duration_sec           int64
is_member              int64
dtype: object
         Unnamed: 0  start_station_code  duration_sec     is_member
count  4.018722e+06        4.018722e+06  4.018722e+06  4.018722e+06
mean   2.009360e+06        6.324815e+03  8.374504e+02  7.992536e-01
std    1.160105e+06        3.758617e+02  6.577147e+02  4.005588e-01
min    0.000000e+00        5.002000e+03  6.100000e+01  0.000000e+00
25%    1.004680e+06        6.105000e+03  3.820000e+02  1.000000e+00
50%    2.009360e+06        6.203000e+03  6.700000e+02  1.000000e+00
75%    3.014041e+06        6.389000e+03  1.121000e+03  1.000000e+00
max    4.018721e+06        1.000200e+04  7.199000e+03  1.000000e+00


In [19]:
#Determine number of missing values for all columns:
c = df_bixi.dtypes.index

for x in c:
    missing=np.where(df_bixi[x].isnull()==True)
    print(x, "column has", len(missing[0]), "missing values")

Unnamed: 0 column has 0 missing values
start_date column has 0 missing values
start_station_code column has 0 missing values
end_date column has 0 missing values
end_station_code column has 0 missing values
duration_sec column has 0 missing values
is_member column has 0 missing values


In [20]:
#size and number of variables in the station csv
print(df_stn.shape)
print(df_stn.head(10))

(546, 5)
   code                                      name   latitude  longitude  \
0  7060                   de l'Église / de Verdun  45.463001 -73.571569   
1  6173                          Berri / Cherrier  45.519088 -73.569509   
2  6203                    Hutchison / Sherbrooke  45.507810 -73.572080   
3  6204                         Milton / Durocher  45.508144 -73.574772   
4  6104                     Wolfe / René-Lévesque  45.516818 -73.554188   
5  6114  Métro Papineau (Cartier / Ste-Catherine)  45.523530 -73.551990   
6  6174                            Roy / St-Denis  45.519080 -73.572700   
7  6719                   Park Row O / Sherbrooke  45.460729 -73.634073   
8  6354                       Marcil / Sherbrooke  45.471743 -73.613924   
9  6223                   du Mont-Royal / du Parc  45.517000 -73.589000   

   is_public  
0          1  
1          1  
2          1  
3          1  
4          1  
5          1  
6          1  
7          1  
8          1  
9          1  


In [21]:
#Let us look at the columns of stations
print(df_stn.dtypes)
#Let us look at details of numeric variables
print(df_stn.describe())

code           int64
name          object
latitude     float64
longitude    float64
is_public      int64
dtype: object
               code    latitude   longitude   is_public
count    546.000000  546.000000  546.000000  546.000000
mean    6412.743590   45.519109  -73.582622    0.983516
std      405.396125    0.027919    0.027344    0.127442
min     5002.000000   45.430740  -73.670634    0.000000
25%     6143.500000   45.501726  -73.599579    1.000000
50%     6305.500000   45.523103  -73.577413    1.000000
75%     6722.750000   45.538784  -73.564608    1.000000
max    10002.000000   45.582757  -73.495067    1.000000


In [22]:
#Determine number of missing values for all columns:
c = df_stn.dtypes.index

for x in c:
    missing=np.where(df_stn[x].isnull()==True)
    print(x, "column has", len(missing[0]), "missing values")

code column has 0 missing values
name column has 0 missing values
latitude column has 0 missing values
longitude column has 0 missing values
is_public column has 0 missing values


# Conclusion:  

For 2017 OD csv file:  
-all datetime columns are defined as an object;  
-end station code is defined as an object;  
-all other columns are defined as an integer;  
-no missing data  

For 2017 station csv file:  
-there are 546 stations  
-for each station, all columns are numeric  
-no missing data